### PR TITLE
Robots.txt

### PR DESCRIPTION
**Pull request by vosmith**
I've added logic to pull and cache robots.txt files for each host the crawler reaches.  Each request verifies the URL against the host's robots.txt file.  This adds a dependency on the https://github.com/temoto/robotstxt project.

A new flag was also added to the `Collector` in order to `IgnoreRobotsTxt`.  Setting this to `true` skips the robots.txt validation.

### CODE REVIEW

File: colly.go [hunk 1]
```diff
@@ -287,6 +299,32 @@ func (c *Collector) isDomainAllowed(domain string) bool {
...
 }
 
+func (c *Collector) checkRobots(u *url.URL) error {
+	var robot *robotstxt.RobotsData
+	var ok bool
+	var err error
+
+	if robot, ok = c.robotsMap[u.Hostname()]; !ok {
+		// no robots file cached
+		resp, _ := c.backend.Client.Get(u.Scheme + "://" + u.Hostname() + "/robots.txt")
```
Comment (asciimoo) [inline colly.go]: `u.Hostname()` -> `u.Host`

File: colly.go [hunk 2]
```diff
@@ -287,6 +299,32 @@ func (c *Collector) isDomainAllowed(domain string) bool {
...
+		robot, err = robotstxt.FromResponse(resp)
+		if err != nil {
+			return err
+		}
+		c.robotsMap[u.Hostname()] = robot
```
Comment (asciimoo) [inline colly.go]: `u.Hostname()` -> `u.Host`

**Comment by vosmith** [comment]
I'll have these changes back to you soon.

File: colly.go [hunk 3]
```diff
@@ -148,6 +154,7 @@ func (c *Collector) Init() {
 	c.backend.Client.CheckRedirect = c.checkRedirectFunc()
 	c.wg = &sync.WaitGroup{}
 	c.lock = &sync.Mutex{}
+	c.robotsMap = make(map[string]*robotstxt.RobotsData, 0)
```
[the remark this answers is not in the document]
Comment (vosmith) [reply colly.go]: I looked into `sync.Map` but i was only available in Go 1.9+, so I opted for adding locks around the writes to `robotsMap` in order to maintain backwards compatibility for previous Go versions

**Comment by asciimoo** [comment]
@vosmith thank you for your fast response. There are still some minor issues, please fix those and the PR is ready to be merged.

**Pull request merged**